### PR TITLE
:sparkles: block move with annotation

### LINKED ISSUE #8473: Allow providers to react when Clusters are paused and indicate in the Cluster's `status` when those actions are finished
### What would you like to be added (User Story)?

As an infrastructure provider developer, I would like to allow my provider to perform and synchronize tasks in reaction to a Cluster being paused instead of assuming that pausing a Cluster happens instantaneously.

### Detailed Description

In CAPZ, we are [planning to adopt Azure Service Operator (ASO)](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3113) to manage individual resources in Azure as Kubernetes objects in place of the Azure SDK from within CAPZ. Since Azure resources are reconciled out-of-band with Cluster API resources in this scenario, CAPZ cannot immediately block ASO's reconciliation loop when a Cluster is `spec.paused`.

ASO reads a separate annotation on each resource it controls to optionally block reconciliation. In response to a Cluster being paused, CAPZ intends to set this annotation on each relevant ASO resource so no reconciliation occurs when a Cluster is paused.  

Since tooling like `clusterctl move` assumes pausing is an instantaneous operation, there is a risk in this case that actions may be taken when the Cluster is not yet fully paused without waiting for some indication that it is in fact paused.

An example of how I envision this might work for CAPZ with `clusterctl move`:

1. `clusterctl move` starts
2. Clusters' `spec.paused` set to `true`
3. `clusterctl move` waits for each Cluster to have `status.paused=true`
4. CAPZ's AzureCluster controller detects its owning Cluster is paused
5. CAPZ annotates all the associated ASO resources to pause reconciliation
6. CAPZ updates the owning Cluster with `status.paused=true`
7. `clusterctl move` continues as it does today

### Anything else you would like to add?

cc @dtzar @CecileRobertMichon 

### Label(s) to be applied

/kind feature
/area api
/area clusterctl

### PR DESCRIPTION
**Pull request by nojnhuh**
**What this PR does / why we need it**:

This PR adds a new well-known `clusterctl.cluster.x-k8s.io/block-move` annotation that, when defined on any resource to be moved, blocks `clusterctl move` from recreating any resources in the target cluster. Infrastructure providers can set this annotation ahead of time to make time to do any extra work in reaction to a cluster being paused.

More details on CAPZ's specific use case are here: #8473.

The expected flow is as follows:
1. Infra controller sets the `block-move` annotation on any resource(s) in the first reconciliation.
1. `clusterctl move` starts.
1. `clusterctl` sets Cluster's `spec.pause` to `true`, waits for `block-move` annotation to not be defined on any resource to be moved.
1. InfraCluster reconciliation invoked based on that change, controller starts work.
1. InfraCluster controller finishes work, removes `block-move`.
1. `clusterctl move` continues.

A POC implementation can be found here which I've verified with one of the self-hosted e2e tests: https://github.com/kubernetes-sigs/cluster-api/compare/main...nojnhuh:cluster-api:wait-pause-docker

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8473

### CODE REVIEW

File: api/v1beta1/common_types.go [hunk 1]
```diff
@@ -91,6 +91,13 @@ const (
 	// on the reconciled object.
 	PausedAnnotation = "cluster.x-k8s.io/paused"
 
+	// IsPausedAnnotation is an annotation whose value indicates whether or not
+	// a Cluster resource is fully paused. It may differ from `spec.paused` when
+	// the Cluster is in the process of pausing or unpausing. This annotation is
+	// only expected to be set by the Cluster API controller, not the user. When
+	// not set, `spec.paused` is assumed to indicate the same status.
+	IsPausedAnnotation = "cluster.x-k8s.io/is-paused"
```
Comment (CecileRobertMichon) [inline api/v1beta1/common_types.go:99]: `cluster.x-k8s.io/is-paused` is a little too close to `cluster.x-k8s.io/paused` and could lead to confusion/human error.  Some alternatives (feel free to suggest more):
- `cluster.x-k8s.io/reconcile-paused`
- `cluster.x-k8s.io/controller-paused`

Alternatively, since this is very much targeting the move use case (? are there other use cases where we see this being used?), this could be something more generic than `paused`:

- `cluster.x-k8s.io/ready-for-move` 
- `cluster.x-k8s.io/paused-to-move-complete`

Comment (killianmuldoon) [reply api/v1beta1/common_types.go:99]: Given that `clusterctl.cluster.x-k8s.io/delete-for-move` was added in https://github.com/kubernetes-sigs/cluster-api/pull/8322 what about someththing like `clusterctl.cluster.x-k8s.io/paused-for-move`?

Comment (nojnhuh) [reply api/v1beta1/common_types.go:99]: I'd vote for something that doesn't include "move" since while that's the main thing we're addressing, this problem exists more generally for anything else that might exist that should wait for clusters to be paused.

**Comment by ykakarap** [comment]
I will take a look at this this week :)

**Comment by fabriziopandini** [comment]
I'm not sure the solution proposed here addresses https://github.com/kubernetes-sigs/cluster-api/issues/8473; I have summarized my understanding of the issue in https://github.com/kubernetes-sigs/cluster-api/issues/8473#issuecomment-1568753318; if possible, let continue discussion on the issue first.

**Comment by nojnhuh** [comment]
Just updated this with something closer to I think what @fabriziopandini had in mind.

/retitle :sparkles: block move with annotation

**Comment by nojnhuh** [comment]
@ykakarap PTAL if you're still interested in going through this.

@fabriziopandini I added some more thoughts to the issue and updated this PR accordingly.

**Comment by nojnhuh** [comment]
I updated the PR description to match how the current changes work.

File: cmd/clusterctl/client/cluster/mover.go [hunk 2]
```diff
@@ -595,6 +610,49 @@ func setClusterClassPause(proxy Proxy, clusterclasses []*node, pause bool, dryRu
...
+
+		if err := retryWithExponentialBackoff(backoff, func() error {
+			if err := c.Get(ctx, key, obj); err != nil {
+				return errors.Wrapf(err, "error getting %s %s", obj.GroupVersionKind(), key)
+			}
+
+			if _, exists := obj.GetAnnotations()[clusterctlv1.MoveBlockedAnnotation]; exists {
+				return errors.Errorf("resource is not ready to move: %s %s", obj.GroupVersionKind(), key)
+			}
+			log.Info("Resource is ready to move", "apiVersion", obj.GroupVersionKind(), "resource", klog.KObj(obj))
```
Comment (CecileRobertMichon) [inline cmd/clusterctl/client/cluster/mover.go:646]: should this be a `v(2)` ? I could see it being noisy for providers that don't use block move if it prints a log line for every single resource

**Comment by nojnhuh** [comment]
(reminder to squash)
/hold

File: cmd/clusterctl/client/cluster/mover.go [hunk 3]
```diff
@@ -595,6 +610,49 @@ func setClusterClassPause(proxy Proxy, clusterclasses []*node, pause bool, dryRu
...
+				Namespace: n.identity.Namespace,
+			},
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: n.identity.APIVersion,
+				Kind:       n.identity.Kind,
+			},
+		}
+		key := client.ObjectKeyFromObject(obj)
+
+		if err := retryWithExponentialBackoff(backoff, func() error {
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover.go:638]: As it is implemented the backoff applies to every CR, thus assuming that we are moving 100 objects, this can result in 100*5sec, which is a lot. Is that the expected behavior?

File: cmd/clusterctl/client/cluster/mover.go [hunk 4]
```diff
@@ -333,6 +335,19 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy, mutators ...Resour
 		return errors.Wrap(err, "error pausing ClusterClasses")
 	}
 
+	log.V(2).Info("Waiting for all resources to be ready to move")
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover.go:338]:
```suggestion
	log.Info("Waiting for all resources to be ready to move")
```
Proposing a higher verbosity so the user gets a stronger signal that the system is waiting for something

File: cmd/clusterctl/client/cluster/mover.go [hunk 5]
```diff
@@ -595,6 +610,49 @@ func setClusterClassPause(proxy Proxy, clusterclasses []*node, pause bool, dryRu
 	return nil
 }
 
+func waitReadyForMove(proxy Proxy, nodes []*node, dryRun bool, backoff wait.Backoff) error {
```
Comment (fabriziopandini) [inline cmd/clusterctl/client/cluster/mover.go:613]: It would be great if we can collect wait for move during discovery, then we can make this a no-op for most of the object, and implement a better UX for objects that are actually blocking move by printing "Wating for ...  to unblock move" before starting to wait

**Review by fabriziopandini — COMMENTED** [review]
Sorry for taking so long to provide feedback.
Overall I think we are really near the finish line, a couple of suggestions from my side but nothing really blocking

Also, could you add the new annotation in https://cluster-api.sigs.k8s.io/reference/labels_and_annotations.html and a note in https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#move; might be worth to surface the new feature to other providers in https://main.cluster-api.sigs.k8s.io/developer/providers/migrations/v1.4-to-v1.5.html

File: cmd/clusterctl/client/cluster/mover.go (hunk 3 above)
Comment (nojnhuh) [reply cmd/clusterctl/client/cluster/mover.go:638]: That's what I had in mind for the time being. I generally wouldn't expect this to ever wait for `N resources * timeout for one resource` since while `clusterctl` is waiting for one resource to be paused, the provider will continue doing work out-of-band to pause that and other resources in parallel. Overall I think the total time this waits will trend toward the longest wait for an individual resource in that case.

That same parallelism could be applied to the waits here as well, but that didn't seem worth the additional complexity under the assumption that resources are going to be paused in parallel.

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/mover.go:638]: I got your point, It is not ideal because ultimately the time the users have to wait can vary in a significant way, but I'm with this for the first iteration if we can get a better UX / like suggested in other comments.

TL;DR;  if there is a block on an object, before starting the wait loop, we should log "Waiting for the block to be removed on object..." or something similar that makes the users aware of where the process is blocked

File: cmd/clusterctl/client/cluster/mover.go (hunk 5 above)
Comment (nojnhuh) [reply cmd/clusterctl/client/cluster/mover.go:613]: Yeah that seems better. I can take a look at that as a follow-up.

**Comment by nojnhuh** [comment]
@fabriziopandini Do you have any other thoughts on this?

**Comment by fabriziopandini** [comment]
@nojnhuh if I'm not wrong this point is still open

> I got your point, but It is not ideal because ultimately the time the users have to wait can vary in a significant way, however, I can be ok with this approach for the first iteration if we can get a better UX as suggested in other comments.

> TL;DR; if (and only if) there is a block on an object, before starting the wait loop we should log "Waiting for the block to be removed on ..." or something similar that makes the users aware of where the move process is blocked at

File: cmd/clusterctl/client/cluster/mover.go (hunk 3 above)
Comment (nojnhuh) [reply cmd/clusterctl/client/cluster/mover.go:638]: Based on my read of the code, it looks like the "resource is not ready to move" error returned here will get logged here which happens before any `time.Sleep`:
https://github.com/kubernetes-sigs/cluster-api/blob/8f173a150434d735409beed2cec58a345926e29a/cmd/clusterctl/client/cluster/client.go#L234

Does that cover what you describe?

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/mover.go:638]: Not really, this is the debug of every retry and it is very noisy; it will also be printed for every object, not only the objects with an actual block move.

So it does not comply with 

> if (and only if) there is a block on an object, before starting the wait loop we should log "Waiting for the block to be removed on ..." or something similar that makes the users aware of where the move process is blocked at

It just provides debug/trace details on every step of the wait loop when you are already in

Comment (nojnhuh) [reply cmd/clusterctl/client/cluster/mover.go:638]: So you'd like to see a single log message indicating that at least one object is blocking the move?

Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/mover.go:638]: obj1 no block move annotation --> no message at all (and possibly skip entirely the wait loop)
obj2 block move annotation --> before starting the wait loop, "Waiting for the block to be removed on ..."
obj3-50 no block move annotation --> no message at all (and possibly skip entirely the wait loop)
obj51 block move annotation --> before starting the wait loop, "Waiting for the block to be removed on ..."
and so on

Comment (nojnhuh) [reply cmd/clusterctl/client/cluster/mover.go:640]: That makes sense, I've updated this to do that. FWIW we need to enter the wait loop no matter what to determine if the annotation exists, but the loop is exited as soon as the annotation doesn't exist (which may be in the first iteration). You already suggested gathering whether the resource is blocked during object discovery which could make this short-circuit earlier, but I'd like to do that as a follow-up if that's still acceptable.

**Comment by CecileRobertMichon** [comment]
/cc @ykakarap @Jont828 

Are you both able to review this to help move it forward?

**Comment by Jont828** [comment]
@CecileRobertMichon @nojnhuh Sure thing, I'll take a look today!

File: docs/book/src/clusterctl/commands/move.md [hunk 6]
```diff
@@ -32,6 +32,8 @@ The discovery mechanism for determining the objects to be moved is in the [provi
 
 Before moving a `Cluster`, clusterctl sets the `Cluster.Spec.Paused` field to `true` stopping
 the controllers from reconciling the workload cluster _in the source management cluster_.
+clusterctl will then wait for the `clusterctl.cluster.x-k8s.io/block-move` annotation not
```
Comment (Jont828) [inline docs/book/src/clusterctl/commands/move.md:35]:
```suggestion
clusterctl will wait until the `clusterctl.cluster.x-k8s.io/block-move` annotation is not present on any resource targeted by the move operation.
```

**Review by Jont828 — COMMENTED** [review]
Nice work! Just some small things

**Comment by Jont828** [comment]
LGTM pending squash.

**Comment by nojnhuh** [comment]
squashed!
/hold cancel

File: cmd/clusterctl/client/cluster/mover.go [hunk 7]
```diff
@@ -333,6 +335,19 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy, mutators ...Resour
 		return errors.Wrap(err, "error pausing ClusterClasses")
 	}
 
+	log.V(2).Info("Waiting for all resources to be ready to move")
+	// exponential backoff configuration which returns durations for a total time of ~2m.
+	// Example: 0, 5s, 8s, 11s, 17s, 26s, 38s, 57s, 86s, 128s
+	waitForMoveUnblockedBackoff := wait.Backoff{
+		Duration: 5 * time.Second,
```
[the remark this answers is not in the document]
Comment (CecileRobertMichon) [reply cmd/clusterctl/client/cluster/mover.go:342]: @nojnhuh could you please open an issue to track this?

File: cmd/clusterctl/client/cluster/mover.go [hunk 8]
```diff
@@ -595,6 +610,55 @@ func setClusterClassPause(proxy Proxy, clusterclasses []*node, pause bool, dryRu
...
+		blockLogged := false
+		if err := retryWithExponentialBackoff(backoff, func() error {
+			if err := c.Get(ctx, key, obj); err != nil {
+				return errors.Wrapf(err, "error getting %s/%s", obj.GroupVersionKind(), key)
+			}
+
+			if _, exists := obj.GetAnnotations()[clusterctlv1.BlockMoveAnnotation]; exists {
+				if !blockLogged {
+					log.Info("Move blocked by %s annotation, waiting for it to be removed")
+					blockLogged = true
```
Comment (CecileRobertMichon) [inline cmd/clusterctl/client/cluster/mover.go:648]: any reason we're only logging the first resource?

Comment (nojnhuh) [reply cmd/clusterctl/client/cluster/mover.go:648]: I was trying to follow what @fabriziopandini was describing here: https://github.com/kubernetes-sigs/cluster-api/pull/8690#discussion_r1258727264

File: cmd/clusterctl/client/cluster/mover.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (nojnhuh) [reply cmd/clusterctl/client/cluster/mover.go:342]: Done: https://github.com/kubernetes-sigs/cluster-api/issues/9194

**Comment by nojnhuh** [comment]
Need to squash again
/hold

File: cmd/clusterctl/client/cluster/mover.go [hunk 9]
```diff
@@ -595,6 +610,55 @@ func setClusterClassPause(proxy Proxy, clusterclasses []*node, pause bool, dryRu
...
+
+		blockLogged := false
+		if err := retryWithExponentialBackoff(backoff, func() error {
+			if err := c.Get(ctx, key, obj); err != nil {
+				return errors.Wrapf(err, "error getting %s/%s", obj.GroupVersionKind(), key)
+			}
+
+			if _, exists := obj.GetAnnotations()[clusterctlv1.BlockMoveAnnotation]; exists {
+				if !blockLogged {
+					log.Info("Move blocked by %s annotation, waiting for it to be removed")
```
[the remark this answers is not in the document]
Comment (nojnhuh) [reply cmd/clusterctl/client/cluster/mover.go]: Yep, fixed

File: cmd/clusterctl/client/cluster/mover.go (hunk 8 above)
Comment (CecileRobertMichon) [reply cmd/clusterctl/client/cluster/mover.go:648]: I see now that I re-read the code more closely it makes more sense. We're not logging the first object, we're logging the first "attempt" for every object.

File: cmd/clusterctl/client/cluster/mover.go (hunk 3 above)
Comment (CecileRobertMichon) [reply cmd/clusterctl/client/cluster/mover.go:640]: > I'd like to do that as a follow-up if that's still acceptable

this would also be good to track in an issue

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/assign @fabriziopandini

File: cmd/clusterctl/client/cluster/mover.go (hunk 3 above)
Comment (nojnhuh) [reply cmd/clusterctl/client/cluster/mover.go:640]: Opened #9196

File: cmd/clusterctl/client/cluster/mover.go [hunk 10]
```diff
@@ -333,6 +335,19 @@ func (o *objectMover) move(graph *objectGraph, toProxy Proxy, mutators ...Resour
 		return errors.Wrap(err, "error pausing ClusterClasses")
 	}
 
+	log.Info("Waiting for all resources to be ready to move")
+	// exponential backoff configuration which returns durations for a total time of ~2m.
+	// Example: 0, 5s, 8s, 11s, 17s, 26s, 38s, 57s, 86s, 128s
```
Comment (Jont828) [inline cmd/clusterctl/client/cluster/mover.go:340]: Just curious, are these the actual numbers it would use or did you pull this from somewhere?

File: docs/book/src/reference/labels_and_annotations.md [hunk 11]
```diff
@@ -24,6 +24,7 @@
...
+| clusterctl.cluster.x-k8s.io/block-move                           | BlockMoveAnnotation if defined on at least one of the objects in scope for a cluster move operation, prevents the operation to start. Providers' controllers are expected to set the annotation on resources that cannot be instantaneously paused, and remove the annotation when the resource has been actually paused.                                                                                                                                                                                                                                   |
```
Comment (Jont828) [inline docs/book/src/reference/labels_and_annotations.md:27]:
```suggestion
| clusterctl.cluster.x-k8s.io/block-move                           | BlockMoveAnnotation prevents the cluster move operation from starting if it is defined on at least one of the objects in scope. Provider controllers are expected to set the annotation on resources that cannot be instantaneously paused and remove the annotation when the resource has been actually paused.                                                                                            |
```
^ Update doc to match the doc comment in the type. I think you'd want to reformat the number of spaces between the `|`, though.

File: cmd/clusterctl/client/cluster/mover.go (hunk 10 above)
Comment (nojnhuh) [reply cmd/clusterctl/client/cluster/mover.go:340]: The jitter makes this inherently random to some degree, but I threw together something to simulate what these numbers could be and this is one example.

File: docs/book/src/reference/labels_and_annotations.md (hunk 11 above)
Comment (nojnhuh) [reply docs/book/src/reference/labels_and_annotations.md]: Updated

**Comment by fabriziopandini** [comment]
/lgtm
/approve

@nojnhuh it would be great to get https://github.com/kubernetes-sigs/cluster-api/issues/9196 implemented before 1.6.0 is cut, because the current implementation is adding an unnecessary extra get for all the objects + the user can wait for x times the timeout for waitMove.
